File: asgate/test/fetch_test.dart
import 'dart:convert';

import 'package:asgate/asgate.dart';
import 'package:asgate/src/fetch.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:test/test.dart';

void main() {
  group('AsgateFetch', () {
    test('returns the full envelope on success', () async {
      final fetch = AsgateFetch(
        client: MockClient((req) async {
          expect(req.url.path, '/api/v1/auth/login');
          return http.Response(
            jsonEncode({
              'status_code': 200,
              'message': 'Login exitoso',
              'data': {'access_token': 'x'},
            }),
            200,
          );
        }),
      );
      final json = await fetch.request(
        'https://auth.example.com/api/v1/auth/login',
        AsgateRequestMethod.post,
        body: {'email': 'a@b.c'},
      );
      expect(json['data']['access_token'], 'x');
    });

    test('maps API error envelope to AsgateApiException', () async {
      final fetch = AsgateFetch(
        client: MockClient(
          (req) async => http.Response(
            jsonEncode({
              'status_code': 401,
              'error': {
                'code': 'CLG001_INVALID_CREDENTIALS',
                'userMessage': 'Credenciales inválidas',
              },
            }),
            401,
          ),
        ),
      );
      await expectLater(
        fetch.request(
          'https://auth.example.com/api/v1/auth/login',
          AsgateRequestMethod.post,
        ),
        throwsA(
          isA<AsgateApiException>()
              .having((e) => e.statusCode, 'statusCode', 401)
              .having((e) => e.code, 'code', 'CLG001_INVALID_CREDENTIALS')
              .having(
                (e) => e.isInvalidCredentials,
                'isInvalidCredentials',
                true,
              ),
        ),
      );
    });

    test('maps 5xx to retryable exception', () async {
      final fetch = AsgateFetch(
        client: MockClient((req) async => http.Response('{}', 503)),
      );
      await expectLater(
        fetch.request(
          'https://auth.example.com/api/v1/auth/refresh',
          AsgateRequestMethod.post,
        ),
        throwsA(isA<AsgateRetryableException>()),
      );
    });

    test('maps network errors to retryable exception', () async {
      final fetch = AsgateFetch(
        client: MockClient((req) async {
          throw http.ClientException('connection refused');
        }),
      );
      await expectLater(
        fetch.request(
          'https://auth.example.com/api/v1/auth/refresh',
          AsgateRequestMethod.post,
        ),
        throwsA(isA<AsgateRetryableException>()),
      );
    });

    test('validation error exposes per-field details', () async {
      final fetch = AsgateFetch(
        client: MockClient(
          (req) async => http.Response(
            jsonEncode({
              'status_code': 400,
              'error': {
                'code': 'CSU001_0900',
                'userMessage': 'Datos inválidos',
                'details': [
                  {
                    'field': 'email',
                    'messages': ['invalid email'],
                  },
                ],
              },
            }),
            400,
          ),
        ),
      );
      try {
        await fetch.request(
          'https://auth.example.com/api/v1/auth/signup',
          AsgateRequestMethod.post,
        );
        fail('expected AsgateApiException');
      } on AsgateApiException catch (e) {
        expect(e.code, 'CSU001_0900');
        expect(e.details, isNotNull);
        expect(e.details!.first.field, 'email');
        expect(e.details!.first.messages, ['invalid email']);
      }
    });
  });
}

File: asgate-js/asgate-js/test/fetch.test.ts
import { describe, expect, it } from 'vitest'
import { AsgateFetch } from '../src/lib/fetch'
import {
  AsgateApiException,
  AsgateRetryableException,
} from '../src/lib/errors'
import { createMockFetch } from './helpers'

const ENVELOPE_OK = {
  status_code: 200,
  message: 'Login exitoso',
  data: { access_token: 'tok', refresh_token: 'ref' },
}

const ERROR_ENVELOPE = {
  status_code: 401,
  error: {
    code: 'CLG001_INVALID_CREDENTIALS',
    userMessage: 'Credenciales inválidas',
  },
}

describe('AsgateFetch', () => {
  it('devuelve el envelope en respuestas 2xx', async () => {
    const { fetchImpl } = createMockFetch(() => ({
      status: 200,
      body: ENVELOPE_OK,
    }))
    const fetch = new AsgateFetch({ fetch: fetchImpl })
    const json = await fetch.request('http://x/api/v1/auth/login', 'post', {
      body: { email: 'a@b.c', password: 'x' },
    })
    expect(json['data']).toEqual({ access_token: 'tok', refresh_token: 'ref' })
  })

  it('solo envía Content-Type cuando hay body', async () => {
    const { fetchImpl, calls } = createMockFetch(() => ({
      status: 200,
      body: ENVELOPE_OK,
    }))
    const fetch = new AsgateFetch({ fetch: fetchImpl })
    await fetch.request('http://x/api/v1/auth/me', 'get')
    const headers = calls[0]?.init.headers as Record<string, string>
    expect(headers['Content-Type']).toBeUndefined()
    expect(headers['Accept']).toBe('application/json')
  })

  it('mapea el envelope de error a AsgateApiException', async () => {
    const { fetchImpl } = createMockFetch(() => ({
      status: 401,
      body: ERROR_ENVELOPE,
    }))
    const fetch = new AsgateFetch({ fetch: fetchImpl })
    await expect(
      fetch.request('http://x/api/v1/auth/login', 'post', {}),
    ).rejects.toMatchObject({
      name: 'AsgateApiException',
      statusCode: 401,
      code: 'CLG001_INVALID_CREDENTIALS',
    })
  })

  it('mapea errores de validación con details', async () => {
    const { fetchImpl } = createMockFetch(() => ({
      status: 400,
      body: {
        status_code: 400,
        error: {
          code: '0900',
          userMessage: 'Validación',
          details: [{ field: 'email', messages: ['Email inválido'] }],
        },
      },
    }))
    const fetch = new AsgateFetch({ fetch: fetchImpl })
    try {
      await fetch.request('http://x/api/v1/auth/signup', 'post', {})
      expect.unreachable()
    } catch (e) {
      expect(e).toBeInstanceOf(AsgateApiException)
      const api = e as AsgateApiException
      expect(api.details?.[0]?.field).toBe('email')
      expect(api.details?.[0]?.messages).toEqual(['Email inválido'])
    }
  })

  it('mapea 5xx a AsgateRetryableException', async () => {
    const { fetchImpl } = createMockFetch(() => ({
      status: 500,
      body: { status_code: 500, message: 'Internal error' },
    }))
    const fetch = new AsgateFetch({ fetch: fetchImpl })
    await expect(fetch.request('http://x/api/v1/auth/me', 'get')).rejects.toBeInstanceOf(
      AsgateRetryableException,
    )
  })

  it('mapea errores de red a AsgateRetryableException', async () => {
    const { fetchImpl } = createMockFetch(() => {
      throw new Error('ECONNREFUSED')
    })
    const fetch = new AsgateFetch({ fetch: fetchImpl })
    await expect(fetch.request('http://x/a', 'get')).rejects.toBeInstanceOf(
      AsgateRetryableException,
    )
  })
})
